### PR TITLE
Simplify program execution and unit testing with Terminal and CommandLine

### PR DESCRIPTION
**Pull request by hakonhall**
Define new API for executing programs, the most significant being:
 - Terminal: Used to create new CommandLine for each command. A TestTerminal
   can be injected for unit testing that makes it easy to set up
   expectancy on commands, and specify the exit code/output.
 - CommandLine: Add arguments, execute, and retrieve output in various forms,
   and much more.

The idea is to remove Command and friends after this is stable. It's not done
in the same PR as multiple repos depends on it.

Having being notified of commons-exec, this is quite similar on the surface. A
few differences:
 - The Terminal-suite makes it much easier to get command logging correct with
   its execute() (=> logging) or executeSilently() (with
   recordSilentExecutionAsSystemModification if it turns out you need to log
   anyway, which is needed for e.g. YUM commands).
 - commons-exec doesn't seem to produce nice error message like I made some
   effort to do correctly: E.g. making a snippet of the output in case an
   exception is thrown. This can be vital to understanding an error.
 - commons-exec requires some discipline on behalf of the developer to get unit
   testable code. He Terminal and CommandLine classes could wrap and hide much
   of commons-exec to make this easier.
 - commons-exec is thread-heavy: using multiple threads for e.g. observing
   timeouts, for shuffling around the stdout, stderr, and stdin data, etc.
 - commons-exec doesn't seem to use SIGKILL in case SIGTERM doesn't stop the
   process.

On the other hand, commons-exec supports some additional features like setting
working directory or environment variables. This will have to be added to
Terminal/CommandLine later.

### CODE REVIEW

File: node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/task/util/process/ProcessFactoryImpl.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        Path temporaryFile = uncheck(() -> Files.createTempFile(
+                temporaryFilePrefix,
+                ".out",
+                fileAttribute));
+
+        try {
+            processBuilder.redirectOutput(temporaryFile.toFile());
+            ProcessApi2 process = processStarter.start(processBuilder);
+            ChildProcess2Impl child = new ChildProcess2Impl(commandLine, process, temporaryFile, timer);
+            // child is now responsible for deleting temporaryFile (it's AutoClosable)
```
Comment (freva) [inline node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/task/util/process/ProcessFactoryImpl.java]: How will `close()` be called? You need to define `child` in the resource scope?

Comment (hakonhall) [reply node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/task/util/process/ProcessFactoryImpl.java]: The child is returned immediately, and the caller has to put it in resource scope.

**Review by smorgrav — APPROVED** [review]
We have also discussed this vs other process implementations such as apache commons exec - this is basically an abstraction over that and it facilitates testing thus I'm good with the 1.5K LOC for this.

File: node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/task/util/process/CommandLine.java [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+        return doExecute();
+    }
+
+    /**
+     * Same as execute(), except it will not record the program as modifying the system.
+     *
+     * If the program is later found to have modified the system, or otherwise worthy of
+     * a record, call recordSilentExecutionAsSystemModification().
+     */
+    public CommandResult executeSilently() {
```
[the remark this answers is not in the document]
Comment (hakonhall) [reply node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/task/util/process/CommandLine.java]: That was considered, however
 - The developer should decide, for every command, whether the command modifies the system or not (silent vs not). While fluent methods is more typically used to amend a call to get additional features you want (add permissions, set limits, etc).
 - Having the same prefix (execute) for both execute and executeSilently is more likely to make the developer aware of both variants (e.g. completion in IntelliJ will show both as you type) and force a correct decision.
 - I think having a single method take a boolean or enum is actually more verbose and/or less clear, e.g. execute(true) is short but not clear, and executeSilently(true) is clear but verbose.

File: node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/task/util/process/TerminalImpl.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
+// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.vespa.hosted.node.admin.task.util.process;
+
+import com.yahoo.vespa.hosted.node.admin.component.TaskContext;
+
+/**
+ * @author hakonhall
+ */
+public class TerminalImpl {
```
[the remark this answers is not in the document]
Comment (hakonhall) [reply node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/task/util/process/TerminalImpl.java]: Fixed in upcomming commit.

File: node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/task/util/process/ProcessFactoryImpl.java (hunk 1 above)
Comment (freva) [reply node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/task/util/process/ProcessFactoryImpl.java]: Ah, right. This `finally` block is confusing me. The idea is that we want to remove the temp file if we fail to create `child`? Why not put this in `catch` and remove the `if (temporaryFile != null)` check?

**Comment by hakonhall** [comment]
PTAL

File: node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/task/util/process/ProcessFactoryImpl.java (hunk 1 above)
Comment (hakonhall) [reply node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/task/util/process/ProcessFactoryImpl.java]: Yes, let me do that.

Comment (hakonhall) [reply node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/task/util/process/ProcessFactoryImpl.java]: Fixed

**Review by freva — APPROVED** [review]

**Pull request merged**